src/wp-includes/script-loader.php

```diff
@@ -2267,38 +2267,51 @@ function wp_enqueue_global_styles() {
 		return;
 	}
 
+	$separate_assets = wp_should_load_separate_core_block_assets();
+
+	// Global styles should be displayed in the header when loading combined.
+	if ( ! $separate_assets && doing_action( 'wp_footer' ) ) {
```

Comment:
What if we bail early for all the cases in which we don't need to register anything?

`! $separate_assets && doing_action( 'wp_footer' ) || $separate_assets && doing_action('wp_enqueue_scripts')`

That way we don't need the other two checks below (line 2284, 2312) and avoid situations in which we register the styles but don't enqueue them. The code would be less nested and simpler to follow as well.

Reply:
This also has the advantage of reiterating that this callback is by default hooked to both `'wp_footer'` and `'wp_enqueue_scripts'`.

Reply:
I'm not sure I follow.

I can speak for how this function for enqueueing global styles should behave: these styles need to be enqueued when the theme has a `theme.json` file. They can be present in the head or footer (only in one place) and that decision depends on the `wp_should_load_separate_core_block_assets` boolean.

Other than that, global styles don't care what the "separate core block assets" logic does. In the future, we may want to have that logic in consideration to only enqueue global styles relevant to the blocks in use, but we need to look at this deeply (not part of WordPress 5.8).

Does this help?

Reply:
It does a bit. I was mainly calling attention to the difference between `wp_register_style()` and `wp_enqueue_style()` and questioning whether styles should always be _registered_ regardless of whether they are enqueued.

For the most part, any style or script that is available in some form is _registered_ by Core. That allows any plugin or theme to use them however they need. But they may or not be _enqueued_ by Core. If these are only meant to be used when separating core block assets is desired on traditional themes, then it's probably safe just leave this as is. Since this is a new "feature" as well, that would be considered an enhancement anyway.

Reply:
Sorry, "leave this as is" here meant leave it as it is in Core currently (returning early according to your suggestion). Updating the PR now.